### PR TITLE
Show status problems in staging project problems column

### PR DESCRIPTION
**Pull request by krauselukas**

### CODE REVIEW

File: src/api/app/models/status/check.rb [hunk 1]
```diff
@@ -10,6 +10,7 @@ class Status::Check < ApplicationRecord
   validates :state, :name, presence: true
   # TODO: This should be an ENUM
   VALID_STATES = %w[pending error failure success].freeze
+  FAILED_STATES = %w[error failure].freeze
```
Comment (coolo) [inline src/api/app/models/status/check.rb]: then better use this in line 47

File: src/api/app/models/staging/staging_project.rb [hunk 2]
```diff
@@ -133,6 +142,14 @@ def build_state
       :acceptable
     end
 
+    def relevant_checks
+      @relevant_checks ||= checks.where(status_reports: { uuid: repositories.map(&:build_id) })
+    end
+
+    def relevant_checks_for_architectures
+      @relevant_checks_for_architectures ||= checks_for_architectures.where(status_reports: { uuid: repository_architectures.map(&:build_id) })
```
Comment (Ana06) [inline src/api/app/models/staging/staging_project.rb]: same here

File: src/api/app/models/staging/staging_project.rb [hunk 3]
```diff
@@ -100,6 +104,11 @@ def unassign_managers_group(managers)
       relationships.find_by(group: managers, role: role).try(:destroy!)
     end
 
+    def failed_status_checks
+      relevant_checks.where(state: Status::Check::FAILED_STATES) +
```
Comment (Ana06) [inline src/api/app/models/staging/staging_project.rb]: [nitpick] @ChrisBr may prefer you use `union` for readability :stuck_out_tongue_closed_eyes:

File: src/api/app/models/staging/staging_project.rb [hunk 4]
```diff
@@ -133,6 +142,14 @@ def build_state
       :acceptable
     end
 
+    def relevant_checks
+      @relevant_checks ||= checks.where(status_reports: { uuid: repositories.map(&:build_id) })
```
Comment (Ana06) [inline src/api/app/models/staging/staging_project.rb]: why do you assign it to a variable? this is only used once.

`pluck(:build_id)` instead of `map`

You should move it to an [scope](https://guides.rubyonrails.org/active_record_querying.html#scopes)

File: src/api/app/models/staging/staging_project.rb (hunk 3 above)
Comment (Ana06) [reply src/api/app/models/staging/staging_project.rb]: this is used more than once, you should cache it

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 5]
```diff
@@ -2,20 +2,31 @@
   max_problems = 5
   project_class = staging_project.name.tr(':', '_')
   item_project_class = 'hidden-item-' + project_class
+  left_problems = max_problems
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: what `left_problems` means?

I don't think this variable is needed...

File: src/api/app/views/webui2/webui/staging/workflows/_problems_item.html.haml [hunk 6]
```diff
@@ -1,16 +0,0 @@
-- staging_project.problems[index].each do |name, states|
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/staging/workflows/_problems_item.html.haml]: why did you rename this file? :confused:

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 7]
```diff
@@ -2,20 +2,31 @@
...
+- if build_problems.empty? && failed_status_checks.empty?
   .text-center
     %i.fas.fa-check-circle.text-success
 - else
   %ul.list-group.list-group-flush
-    = render partial: 'problems_item', locals: { staging_project: staging_project, index: 0...max_problems }
-
-    - length = staging_project.problems.length
-    - if length > max_problems
+    - failed_status_checks[0...max_problems].each do |check|
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: do we want to render openQA problems first? :thinking:

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 8]
```diff
@@ -2,20 +2,31 @@
...
+      = render partial: 'status_problems', locals: { check: check }
+      - failed_status_checks.shift
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: why do you need this? :unamused:

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 9]
```diff
@@ -2,20 +2,31 @@
...
+      - left_problems -= 1
+    - if left_problems > 0
+      - build_problems[0...left_problems].each do |name, states|
+        = render partial: 'build_problems', locals: { name: name, states: states, staging_project: staging_project }
+        - build_problems.shift
```
Comment (Ana06) [inline src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: why do you need this?

File: src/api/app/models/staging/staging_project.rb (hunk 4 above)
Comment (ChrisBr) [reply src/api/app/models/staging/staging_project.rb]: ``build_id`` is not a field, it is coming from the backend, you can not pluck it.

File: src/api/app/models/staging/staging_project.rb (hunk 2 above)
Comment (ChrisBr) [reply src/api/app/models/staging/staging_project.rb]: build_id is not a field, it is coming from the backend, you can not pluck it.

File: src/api/app/views/webui2/webui/staging/workflows/_problems_item.html.haml (hunk 6 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems_item.html.haml]: @Ana06 Since we render to partials now, i thought its more specific to call it build_problems and status_problems instead of just problems_item

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 7 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @Ana06 not sure, at least in the current obs_factory staging they are shown first, so i took it over

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 8 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @ana06 i guess you are talking about the ```shift```, i first tried it without but it ended up using helper variables to indicate where to continue the iteration in the hidden part of the list, so i thought it easier to just remove the rendered items and just rendering the left overs in the collapsed part

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 9 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @Ana06 i guess you are talking about the shift, i first tried it without but it ended up using helper variables to indicate where to continue the iteration in the hidden part of the list, so i thought it easier to just remove the rendered items and just rendering the left overs in the collapsed part

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 5 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @ana06 i will check if i can do it without, otherwise i will make the name a bit more meaningful :)

**Comment by krauselukas** [comment]
@ana06 yeah sure, i will add an screenshot

**Comment by krauselukas** [comment]
@Ana06 Screenshots :tada: 

screenshot-2018-11-21 staging for foo - open build service
screenshot-2018-11-21 staging for foo - open build service 1

File: src/api/app/models/staging/staging_project.rb (hunk 3 above)
Comment (krauselukas) [reply src/api/app/models/staging/staging_project.rb]: @Ana06 caching: done

File: src/api/app/models/status/check.rb (hunk 1 above)
Comment (krauselukas) [reply src/api/app/models/status/check.rb]: @coolo done

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 5 above)
Comment (krauselukas) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @Ana06 check again, i got rid of the variable

**Comment by krauselukas** [comment]
@ana06 i added a commit description and fixed the typo

File: src/api/app/models/staging/staging_project.rb (hunk 2 above)
Comment (Ana06) [reply src/api/app/models/staging/staging_project.rb]: It is not the first time I make this comment :see_no_evil:

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml (hunk 8 above)
Comment (Ana06) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: you can do this with the length, this variable is confusing and not needed

File: src/api/app/views/webui2/webui/staging/workflows/_status_problems.html.haml [hunk 10]
```diff
@@ -0,0 +1,3 @@
+%li.list-group-item.table-list-group-item
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/views/webui2/webui/staging/workflows/_status_problems.html.haml]: why not? the alternative is to repeat it

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 11]
```diff
@@ -1,21 +1,29 @@
 :ruby
-  max_problems = 5
+  max_shown = 5
   project_class = staging_project.name.tr(':', '_')
   item_project_class = 'hidden-item-' + project_class
+  checks_problems = staging_project.failed_status_checks
+  build_problems = staging_project.problems
+  total_problems = checks_problems.length + build_problems.length
+  rendered_checks_problems = [max_shown, checks_problems.length].min
+  rendered_build_problems = max_shown - rendered_checks_problems
```
Comment (krauselukas) [inline src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @ana06 i know its better to assign them after the else statement but rubocop was complaining :(

**Comment by bgeuken** [comment]
@krauselukas btw. please always add the screenshots to the initial comment. Otherwise they are easily overlooked when the number of comments grows.

File: src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml [hunk 12]
```diff
@@ -1,21 +1,29 @@
...
+- if total_problems == 0
   .text-center
     %i.fas.fa-check-circle.text-success
 - else
   %ul.list-group.list-group-flush
-    = render partial: 'problems_item', locals: { staging_project: staging_project, index: 0...max_problems }
-
-    - length = staging_project.problems.length
-    - if length > max_problems
+    - checks_problems[0..(rendered_checks_problems - 1)].each do |check|
```
[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @bgeuken 

> Or use limit and offset if this is an active record query object.

it is not

[the remark this answers is not in the document]
Comment (Ana06) [reply src/api/app/views/webui2/webui/staging/workflows/_problems.html.haml]: @bgeuken 

> You could write `checks_problems[0...rendered_checks_problems]`

you could but I would keep all the indexes consistent and if you use `...` everywhere then you will have `-1` in others :unamused:

**Comment by coolo** [comment]
I removed the broken required reports - so we're green

File: src/api/app/models/staging/staging_project.rb [hunk 13]
```diff
@@ -112,6 +112,11 @@ def current_missing_checks
       (relevant_status_reports + relevant_status_reports_for_architectures).map(&:missing_checks).flatten
     end
 
+    def failed_status_checks
+      @failed_status_checks ||= relevant_checks.where(state: Status::Check::FAILED_STATES) +
```
Comment (coolo) [inline src/api/app/models/staging/staging_project.rb]: so @ChrisBr there is no merge conflict per se, but relevant_checks no longer exists

Comment (ChrisBr) [reply src/api/app/models/staging/staging_project.rb]: Updated

Comment (coolo) [reply src/api/app/models/staging/staging_project.rb]: but not pushed? :)
